### ISSUE TITLE
Validate task-graph.json against schema

### ISSUE THREAD

**Issue by mitchhentges** [mozilla-releng/scriptworker#355]
Related to #353: if an invalid `task-graph.json` is provided, we should be able to detect it earlier in the CoT process and provide a more-useful error, like `malformed-payload` (rather than a `KeyError` deep in CoT).

**Comment by escapewindow** [comment]
This check probably belongs in taskgraph.

**Comment by mitchhentges** [comment]
:thinking: However, `scriptworker` is a self-contained unit, and input isn't guaranteed to come from `taskgraph`. A schema is meant to enforce requirements on the input that a service processes - imagine when `scriptworker` is updated to handle new features and requires new properties. It's more straightforward to keep the schema in lockstep with `scriptworker`, rather than being forced to ensure that every `taskgraph` generating graphs has an updated schema before we can update `scriptworker`

**Comment by escapewindow** [comment]
If it were in taskgraph, we'd get the error at the end of the decision task, rather than at the first scriptworker run.

**Comment by mitchhentges** [comment]
It's true that doing schema validation at the end of a decision task run, if there were issues, they'd be caught sooner.
However, each `scriptworker` is like a standalone service with an API (the payload, incoming artifacts, etc), and **`taskgraph` is not the only possible consumer.**

API validation gives you two things:
1. Allows you to confidently write code in `scriptworker` that can safely assume that certain properties exist and certain requirements are met. If validation is done at the end of `taskgraph`, you have to write `scriptworker` code much more defensively.
2. Gives a heads up to the caller if something is set wrong

Putting the `schema` in `taskgraph` would only resolve the 2nd point.

**Comment by escapewindow** [comment]
Sure. We can put the schema check in both places, but I believe the plan is for taskgraph to become the only, or primary, supported consumer.

**Comment by mitchhentges** [comment]
I agree that `taskgraph` will be the primary supported consumer, but since we can't _enforce_ that that's the case, I believe that we'll run into less surprises if we validate within `scriptworker`.

I'm imagining it like a client and server - you might have a client that's used 99% of the time (`taskgraph`), but it's probably worth having validation on the server so you don't run into any surprises

**Comment by escapewindow** [comment]
👍 

If we find in practice that we catch all the errors in taskgraph and have no real practical use case for the schema check in scriptworker, we can remove it or make it optional at that point.

**Comment by escapewindow** [comment]
I think it's too complex to have the check in scriptworker: we'd have to add schemas for gecko, various flavors of mobile, xpi, mpd, and any other taskgraph repo we end up supporting, and these will quickly go out of date or be invalid for esr when they're valid for central. We should add tests in taskgraph as we find them.

Also, we have decided to standardize on taskgraph, so supporting non-taskgraph solutions in scriptworker is not currently on the roadmap. Closing.
